File: pkg/ios/GupSwift/Sources/GupSwift/GupBridge.swift
// Copyright (C) 2024 Corin Lawson
// SPDX-License-Identifier: GPL-3.0-or-later

/// C-ABI bridge to the Rust `gup-ios` static library.
///
/// These declarations must stay in sync with `gup-ios/src/lib.rs`.
/// The static library (`libgup_ios.a`) is linked via Xcode build settings.

import Foundation

// MARK: - Opaque Handle

/// Opaque pointer to the Rust-side `GupContext`.
public typealias GupContextHandle = OpaquePointer?

// MARK: - C ABI Declarations

/// Create a new Gup GPU context. Returns `nil` on failure.
@_silgen_name("gup_context_create")
public func gup_context_create() -> GupContextHandle

/// Destroy a context previously created with `gup_context_create`.
@_silgen_name("gup_context_destroy")
public func gup_context_destroy(_ handle: GupContextHandle)

/// Attach a CAMetalLayer-backed UIView to the context.
/// Returns a surface ID (> 0) on success, 0 on failure.
@_silgen_name("gup_surface_attach_layer")
public func gup_surface_attach_layer(
    _ handle: GupContextHandle,
    _ uiView: UnsafeMutableRawPointer,
    _ uiViewController: UnsafeMutableRawPointer?,
    _ width: UInt32,
    _ height: UInt32
) -> UInt64

/// Detach a surface from the context.
@_silgen_name("gup_surface_detach")
public func gup_surface_detach(_ handle: GupContextHandle, _ surfaceId: UInt64)

/// Render one frame to the given surface.
@_silgen_name("gup_render_frame")
public func gup_render_frame(_ handle: GupContextHandle, _ surfaceId: UInt64) -> Bool

/// Forward a UITouch event into the Gup event pipeline.
@_silgen_name("gup_touch_event")
public func gup_touch_event(
    _ handle: GupContextHandle,
    _ touchId: UInt64,
    _ x: Float,
    _ y: Float,
    _ phase: UInt8,
    _ scaleFactor: Float,
    _ timestamp: Double,
    _ viewWidth: Float,
    _ viewHeight: Float
)

/// Notify Gup of a drawable-size change (orientation change).
@_silgen_name("gup_surface_resize")
public func gup_surface_resize(
    _ handle: GupContextHandle,
    _ surfaceId: UInt64,
    _ newWidth: UInt32,
    _ newHeight: UInt32
) -> Bool

File: pkg/android/GupKotlin/src/main/java/au/com/phiware/gup/GupBridge.kt
// Copyright (C) 2024 Corin Lawson
// SPDX-License-Identifier: GPL-3.0-or-later

package au.com.phiware.gup

/**
 * JNI bridge to the `gup-android` native library.
 *
 * All functions map 1:1 to the `extern "C"` functions exported by
 * `gup-android/src/lib.rs`.  The Kotlin-side wrappers ([GupContext],
 * [GupSurfaceView]) provide the ergonomic, lifecycle-aware API.
 */
internal object GupBridge {
    init {
        System.loadLibrary("gup_android")
    }

    // -- Context lifecycle ----------------------------------------------------

    /** Create a new GPU context.  Returns an opaque native handle (pointer). */
    external fun nativeCreate(): Long

    /** Destroy the GPU context. */
    external fun nativeDestroy(handle: Long)

    // -- Surface lifecycle (SurfaceHolder.Callback) ---------------------------

    /** Attach an ANativeWindow and return a surface ID (> 0), or 0 on error. */
    external fun nativeSurfaceCreated(handle: Long, surface: android.view.Surface, width: Int, height: Int): Long

    /** Notify the native side of a surface resize. */
    external fun nativeSurfaceChanged(handle: Long, surfaceId: Long, width: Int, height: Int): Boolean

    /** Detach the surface and release GPU resources. */
    external fun nativeSurfaceDestroyed(handle: Long, surfaceId: Long)

    // -- Rendering ------------------------------------------------------------

    /** Render a single frame.  Returns true on success. */
    external fun nativeRenderFrame(handle: Long, surfaceId: Long): Boolean

    // -- Touch events ---------------------------------------------------------

    /** Forward a single MotionEvent pointer into the Gup event pipeline. */
    external fun nativeOnTouchEvent(
        handle: Long,
        pointerId: Long,
        x: Float,
        y: Float,
        action: Int,
        density: Float,
        eventTimeMs: Double,
        viewWidth: Float,
        viewHeight: Float
    )

    // -- Activity lifecycle ---------------------------------------------------

    /** Notify the native side that the Activity is pausing. */
    external fun nativePause(handle: Long)

    /** Notify the native side that the Activity is resuming. */
    external fun nativeResume(handle: Long)
}
